### ISSUE TITLE
Rust cargo linter fails

### ISSUE THREAD

**Issue by crisidev** [w0rp/ale#1045]
I am using rust stable and the cargo based linter fails without errors or output. Here is the ALEInfo output

```

 Current Filetype: rust
Available Linters: ['cargo', 'rls', 'rustc']
  Enabled Linters: ['cargo']
 Linter Variables:

let g:ale_rust_cargo_use_check = 0
 Global Variables:

let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%s'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {}
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 750
let g:ale_lint_on_enter = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'always'
let g:ale_linter_aliases = {}
let g:ale_linters = {'vim': ['vint'], 'go': ['gometalinter'], 'markdown': ['mdl'], 'sh': ['shellcheck'], 'html': ['tidy'], 'ruby': ['rubocop'], 'rust': ['cargo'], 'python': ['flake8', 'mypy']}
let g:ale_open_list = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '✗'
let g:ale_sign_offset = 1000000
let g:ale_sign_warning = '•'
let g:ale_statusline_format = ['%d error(s)', '%d warning(s)', 'OK']
let g:ale_warn_about_trailing_whitespace = 1
  Command History:

(executable check - success) cargo
(started) ['/bin/zsh', '-c', 'cargo build --frozen --message-format=json -q']
(started) ['/bin/zsh', '-c', 'cargo build --frozen --message-format=json -q']
(finished - exit code 101) ['/bin/zsh', '-c', 'cargo build --frozen --message-format=json -q']

<<<NO OUTPUT RETURNED>>>
```

If I run the command manually in my source directory, it works fine, printing all the warnings in JSON:

```
/bin/zsh -c "cargo build --frozen --message-format=json -q"                                                   
{"message":{"children":[{"children":[],"code":null,"level":"note","message":"#[warn(dead_code)] on by default","rendered":null,"spans":[]}],"code":null,"level":"warning","message":"struct is never used: `Djb64`","rendered":null,"spans":[{"byte_end":35,"byte_start":0,"column_end":2,"column_start":1,"expansion":null,"file_name":"src/hashmap/mod.rs","is_primary":true,"label":null,"line_end":3,"line_start":1,"suggested_replacement":null,"text":[{"highlight_end":19,"highlight_start":1,"text":"pub struct Djb64 {"},{"highlight_end":15,"highlight_start":1,"text":"    hash: u64,"},{"highlight_end":2,"highlight_start":1,"text":"}"}]}]},"package_id":"hashmap 0.1.0 (path+file:///Users/bigo/Crisidev/software/rusty/hashmap)","reason":"compiler-message","target":{"crate_types":["lib"],"kind":["lib"],"name":"hashmap","src_path":"/Users/bigo/Crisidev/software/rusty/hashmap/src/lib.rs"}}
```

**Comment by aurieh** [comment]
My guess is that you are trying to run cargo outside a cargo package. Could you please make sure that's not the case?

**Comment by crisidev** [comment]
It should be not the case. I tried to open the source file either from the directory where I have Cargo.toml and even from the inner directory where I have the actual source file.

**Comment by crisidev** [comment]
Also this function https://github.com/w0rp/ale/blob/master/ale_linters/rust/cargo.vim#L6 should return empty string if no cargo project has been found..

**Comment by crisidev** [comment]
Another interesting finding: if I just open the rust source file and wait for ale autolint to kick off, it will mark warnings and errors:

```
(executable check - success) cargo
(started) ['/bin/zsh', '-c', 'cargo build --frozen --message-format=json -q']
(finished - exit code 0) ['/bin/zsh', '-c', 'cargo build --frozen --message-format=json -q']
<<<OUTPUT STARTS>>>
{"message":{"children":[{"children":[],"code":null,"level":"note","message":"#[warn(dead_code)] on by default","rendered":null,"spans":[]}],"code":null,"level
":"warning","message":"struct is never used: `Djb64`","rendered":null,"spans":[{"byte_end":35,"byte_start":0,"column_end":2,"column_start":1,"expansion":null,
"file_name":"src/hashmap/mod.rs","is_primary":true,"label":null,"line_end":3,"line_start":1,"suggested_replacement":null,"text":[{"highlight_end":19,"highligh
t_start":1,"text":"pub struct Djb64 {"},{"highlight_end":15,"highlight_start":1,"text":"    hash: u64,"},{"highlight_end":2,"highlight_start":1,"text":"}"}]}]
},"package_id":"hashmap 0.1.0 (path+file:///Users/bigo/Crisidev/software/rusty/hashmap)","reason":"compiler-message","target":{"crate_types":["lib"],"kind
":["lib"],"name":"hashmap","src_path":"/Users/bigo/Crisidev/software/rusty/hashmap/src/lib.rs"}}
...
```
After this first success, every edit or write will end up in the error above.

**Comment by aurieh** [comment]
Oh, that's weird. I have a feeling ale is trying to lint the file outside it's source directory.
I'll see if I can repro this.

**Comment by w0rp** [comment]
You could try setting `g:ale_rust_cargo_use_check` to `1`. That might help, or it might not.

**Comment by w0rp** [comment]
Check the output of `which cargo` in your terminal and `:!which cargo` in Vim. You might need to set `PATH` differently.

**Comment by aurieh** [comment]
@w0rp I've figured it out: recent versions of Cargo now pipe to standard error instead of standard output.
I would PR this fix, but it'll break older versions of cargo.

**Comment by w0rp** [comment]
@aurieh Could you try using `'output_stream': 'both'`, and see if that works? If that works, then we can use that for all versions.

**Comment by w0rp** [comment]
The tests let me know it's pretty safe to do. I have set that now. That should fix the problem.

It looks like the output stream was changed last year on their master branch, so maybe it finally started appearing in stable versions now.

**Comment by crisidev** [comment]
I just updated the repo and it is behaving as before for me:
* Upon file open, ale autochecks the file and I can see the warnings / errors
* If I write the file or start typing, everything breaks and I start seeing this line in ALEInfo:
```
(finished - exit code 101) ['/bin/zsh', '-c', 'cargo build --frozen --message-format=json -q']
<<<NO OUTPUT RETURNED>>>
```

I have last ale commit (0702cb5) and cargo 0.22.0 (3423351a5 2017-10-06)

**Comment by crisidev** [comment]
Both ```which cargo``` from by shell and ```!which cargo``` from vim returns the same path: ```/Users/bigo/.bin/cargo```

**Comment by w0rp** [comment]
>everything breaks 

What do you mean by that?

**Comment by w0rp** [comment]
If you have an error message, please paste it here.

**Comment by w0rp** [comment]
You might also want to try setting `g:ale_rust_cargo_use_check` to `1`.

**Comment by crisidev** [comment]
Sorry, with everything breaks I mean the linting does not work.

I tried to add ```g:ale_rust_cargo_use_check``` to ```1``` and it still doing the same:
* Upon file open, ale autochecks the file and I can see the warnings / errors
* If I write the file or start typing, the linting is not working anymore I start seeing this line in ALEInfo:
```
(finished - exit code 101) ['/bin/zsh', '-c', 'cargo build --frozen --message-format=json -q']
<<<NO OUTPUT RETURNED>>>
```

**Comment by w0rp** [comment]
Oh, is your problem that you don't see any errors _while you type_? You can't see errors while you type with `cargo`, you have to save your file to see some errors.

**Comment by w0rp** [comment]
If you want errors while you type, consider using `rls` instead.

**Comment by w0rp** [comment]
That behaviour is documented in part in the README, and also in `:help ale-rust-options`.

**Comment by w0rp** [comment]
Then again, you say it works when you open a file, but not when you save it? I dunno then.

**Comment by w0rp** [comment]
Share some way of repeating the bug if you can. Maybe a project and a cargo version which together exhibit the same problem.

**Comment by crisidev** [comment]
@w0rp, thanks for pointing out to the documentation. I am not switching to the lint-as-you-type with rls since it looks like requires rust nightly. 

Regarding your question, yes it works when I open a file and never do when I write it.

**Comment by hiberno** [comment]
I can reproduce this issue. 

For what it's worth, here's the output of `ALEInfo`:

```
 Current Filetype: rust
Available Linters: ['cargo', 'rls', 'rustc']
  Enabled Linters: ['cargo']
 Linter Variables:
let g:ale_rust_cargo_check_all_targets = 1
let g:ale_rust_cargo_use_check = 1
 Global Variables:
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%code: %%s'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {}
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'always'
let g:ale_linter_aliases = {}
let g:ale_linters = {}
let g:ale_open_list = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '>>'
let g:ale_sign_offset = 1000000
let g:ale_sign_warning = '--'
let g:ale_statusline_format = ['%d error(s)', '%d warning(s)', 'OK']
let g:ale_warn_about_trailing_whitespace = 1
  Command History:
(executable check - success) cargo
(started) ['/bin/zsh', '-c', 'cargo --version']
(finished - exit code 0) ['/bin/zsh', '-c', 'cargo --version']
(finished - exit code 101) ['/bin/zsh', '-c', 'cargo check --all-targets --frozen --message-format=json -q']
<<<NO OUTPUT RETURNED>>>
(finished - exit code 101) ['/bin/zsh', '-c', 'cargo check --all-targets --frozen --message-format=json -q']
<<<NO OUTPUT RETURNED>>>
(finished - exit code 101) ['/bin/zsh', '-c', 'cargo check --all-targets --frozen --message-format=json -q']
<<<NO OUTPUT RETURNED>>>
```

I wonder if it has something to do with this:

```sh
$ cargo check --all-targets
error: no library targets found
```

Here's the `cargo`-related bug report: https://github.com/rust-lang/cargo/pull/4685 The cargo version I am using is 

```sh
$ cargo --version
cargo 0.23.0 (61fa02415 2017-11-22)
```

It does not work at all when creating a new project running `cargo new --bin` and works only after opening the file when creating a new project like this: `cargo new`.

**Comment by w0rp** [comment]
Please open a new issue, or people might not see the bug report.

**Comment by hiberno** [comment]
@w0rp Sure, sorry. [Here](https://github.com/w0rp/ale/issues/1202) you go.
